### PR TITLE
Add Optional implementation and tests

### PR DESCRIPTION
**Pull request by thenaterhood**

### CODE REVIEW

**Comment by boxcla** [comment]
Hi @thenaterhood, thanks for the pull request. Before we can merge it, we need you to sign our Contributor License Agreement. You can do so electronically here: http://opensource.box.com/cla

Once you have signed, just add a comment to this pull request saying, "CLA signed". Thanks!

File: src/Bart/Optional/Absent.php [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        return null;
+    }
+
+    /**
+     * @param callable $callable
+     * @return mixed
+     */
+    public function map(Callable $callable)
+    {
+        return Optional::absent();
```
Comment (asheepapart) [inline src/Bart/Optional/Absent.php]: Why not `self::instance()`?

File: src/Bart/Optional/Present.php [hunk 2]
```diff
@@ -0,0 +1,102 @@
+<?php
+
+namespace Bart\Optional;
+
+/**
+ * Class that handles containing an existent (non-null) value
+ * Class Present
+ * @package Bart\Optional
+ */
+class Present extends Optional
```
Comment (asheepapart) [inline src/Bart/Optional/Present.php]: Why not call this `Some`?

File: src/Bart/Optional/Absent.php [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+use Bart\Exceptions\IllegalStateException;
+
+/**
+ * Optional package class that handles a nonexistent value. This
+ * class operates as a singleton.
+ * Class Absent
+ * @package Bart\Optional
+ */
+class Absent extends Optional
```
Comment (asheepapart) [inline src/Bart/Optional/Absent.php]: Why not call this `None`?

**Comment by asheepapart** [comment]
Thanks for getting this started!! 

Three high level things:
1. Please give credit to the original author(s) or source(s) of inspiration
2. See comments on methods/classes for the [Scala Option implementation](https://github.com/scala/scala/blob/2.11.x/src/library/scala/Option.scala) and see if you can apply them here. The methods/classes are lacking documentation in several places
3. Each file should have a blank line at the end. This makes git histories much easier to read.

File: src/Bart/Optional/Present.php (hunk 2 above)
Comment (thenaterhood) [inline src/Bart/Optional/Present.php]: I felt like Present was a better name, since either the value is there or it isn't. "Some" would also work, it just seemed to make what was going on a little less clear when reading through the code.

File: src/Bart/Optional/Absent.php (hunk 3 above)
Comment (thenaterhood) [inline src/Bart/Optional/Absent.php]: Similar reasons as Present instead of Some. The other issue I felt with calling this None is that we're avoiding using null with this and None seemed a little too similar. If we're avoiding nulls, then it seemed less ideal to say a lack of a value was an instance of None than it was to say it was Absent.

File: src/Bart/Optional/Absent.php (hunk 1 above)
Comment (thenaterhood) [inline src/Bart/Optional/Absent.php]: I could do that or I could even just return self here, does the same thing. Just a matter of preference in this case, I felt this was a more explicit way of doing it since we're not concerned about having an extra function call.

**Comment by thenaterhood** [comment]
@asheepapart Tried to address your comments.

Also added some conditions and tests for some corner cases so that calling Present->map() using a callable that returns Absent or constructing a Present with an Absent doesn't cause you to end up with Present instances that are holding Absent. Also makes it so you can't have a Present containing a Present (if you try, you end up with a Present that contains the same value as the other one, like a copy constructor).
